proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,626 @@
+by the dapp developer. Penalizes everyone for a few bad actors.
+
+* Passing fees for each account by instruction and validating them
+  during the execution inside user application code
+
+Pros: High efficiency, minimal performance impact, more generic, no need to
+change the account structure.
+
+Cons: Cannot prevent denial of service attacks, or account blocking
+attacks that do not call into the respective program
+
+## New Terminology
+
+Program Rebatable Account Write Fees or PRAW Fees: Fees that are decided by the
+dapp developer will be charged if a user wants to use the dapp. They will be
+applied even if the transaction fails contrary to lamport transfers. The program
+can decide to rebate these fees back to the user if certain conditions decided
+by dapp developers are met.
+
+### Other terminology
+
+Account `Authority`: The true authority for the account. Let's take an example
+of associated token account. A associated token account is a `PDA` derived from
+associated token program, where the owner is token program. But the program
+internally saves who has the authority over the account. Only token program can
+change the data of a token account. For operations like withdrawing tokens, the
+authority of the associated token account has to be the signer for the
+transaction containing transfer instruction. To receive tokens from another
+account, the authority's signature is not required.
+
+`Other Fees`: Fees other than PRAW fees `base fees + prioritization fees`.
+
+## Detailed Design
+
+PRAW fees enable dapp developer to decide who to rebate and who to penalize
+through runtime features. In particular, programs needs to be able to penalize,
+even if the application is not CPI'd to.
+
+The checks on the PRAW fees will be taken care of by Solana runtime. Total PRAW
+fees committed will be included in the transaction through a special parameter
+called `SetProgramRebatableAccountWriteFees` so that it is easier to calculate
+the total amount of fees that will be paid by the transaction. This parameter
+will be part of a new `TransactionHeaderProgram` where all the transaction
+parameters will be set through a TVL. In case of nonpayment or partial payment,
+the program is never executed. This instruction cannot be CPI'ed into.
+
+Addition of new syscalls to get, set and rebate PRAW fees during program
+runtime. Application developers will be able to rebate these fees through the
+new syscall, but only if the transaction succeeds, this way transactions
+write-locking the respective accounts will require full fee payment before
+starting the execution of the transaction. Rebates will be issued to the payer
+after the transaction is executed. During transaction execution, the rebates
+will not be reflected in the payer's balance.
+
+Once PRAW fees are paid on an account by the payer they are valid for the whole
+transaction, even if the same account is write-locked in multiple instructions.
+
```

Comment:
dunno this topic is discussed yet or not. but, 

the existence of write locks of PRAW-fee-enabled accounts must increase CUs even with txes with no ix at all, proportional to the number of those accounts. Otherwise, it works as an obscure lamports transfer mechanism, which cost nothing other than base tx fee. so, this creates security/economic problems..

so, at least, it must consume CU larger than SystemProgram::transfer. (currently, this ix consumes no CU; but planned to be changed).

So add `### Consumption of CUs` subsection under this `## Detailed Design`?

Reply:
So how much CU is a transfer worth? It's fetching and adding of two numbers and storing the result later.

We already have CU for write locks which will cover fetching and storing the result later. So the transfer is just the addition of two numbers.
I see that default cost for system program is 150 CU in the code. So cost of transfer instruction should be also 150 CUs.

Similarly, loading of PRAW fees can come under write-locking of the account. Checks on PRAW fees should be pretty insignificant. `Set` and `Get` PRAW fees should also be insignificant. I guess only `Rebate` remains.

So 150 CUs per account with PRAW fees + 200 CUs per rebate should do the trick.

I would like opinion of @taozhu-chicago .

Reply:
> So how much CU is a transfer worth? It's fetching and adding of two numbers and storing the result later.
> 
> We already have CU for write locks which will cover fetching and storing the result later. So the transfer is just the addition of two numbers. I see that default cost for system program is 150 CU in the code. So cost of transfer instruction should be also 150 CUs.
> 
> Similarly, loading of PRAW fees can come under write-locking of the account. Checks on PRAW fees should be pretty insignificant.

I know the overall computation is low. But, these additional loads/stores incurs some cycles and pollutes cpu cache in the very perf. sensitive code path. Also, enlarges AccountsIndex. imo, i want to solana to squeeze out every ns. :)

Also, I think every additional use of runtime features should incur additional CUs in general. Otherwise it creates incentive skew, at least theoretically.

Assuming PRAW-enabled accounts is very small, well-optimized impl could employ different code path/data. On the other hand, cost-saving validator could just ignore PRAW-enabled txes altogether when producing blocks.
So, PRAW-enabled txes must pay a bit more than other txes with all others being equal.

> So 150 CUs per account with PRAW fees + 200 CUs per rebate should do the trick.

Lastly, 200 CU for the rebate syscall isn't relevant for the case where PRAW is abused to transfer sol.

Reply:
I think it is a technical detail and can be discussed while reviewing the PRs.

Reply:
This is a fair point, accounts loading cost should vary by several factors (eg lock type, account size, account type; #29582 tracks some of these), having PRAW fee can be additional factor. Maybe we can move this discussion to separate thread?

Reply:
Yes I agree, I can add a section like suggested. Will add details what are CU set for ComputeBudget program, and System Program. What should be CU used for TransactionHeader program and each of the syscall. Will summarize a little bit about 29582 and so CU used are subjected to change during its implementation. 

Would it work for you?